plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryClient.java

```diff
@@ -130,34 +143,26 @@ public Optional<RemoteDatabaseObject> toRemoteDataset(String projectId, String d
             return Optional.of(RemoteDatabaseObject.of(datasetName));
         }
 
-        Map<String, Optional<RemoteDatabaseObject>> mapping = new HashMap<>();
-        for (DatasetId datasetId : listDatasetIds(projectId)) {
-            mapping.merge(
-                    datasetId.getDataset().toLowerCase(ENGLISH),
-                    Optional.of(RemoteDatabaseObject.of(datasetId.getDataset())),
-                    (currentValue, collision) -> currentValue.map(current -> current.registerCollision(collision.get().getOnlyRemoteName())));
+        try {
+            Map<String, RemoteDatabaseObject> datasetMap = remoteDatasetCaseInsentiveCache.get(projectId, () -> {
+                Map<String, RemoteDatabaseObject> mapping = new HashMap<>();
+                for (DatasetId dataset : listDatasetIds(projectId)) {
+                    mapping.merge(
+                            dataset.getDataset().toLowerCase(ENGLISH),
+                            RemoteDatabaseObject.of(dataset.getDataset()),
+                            (currentValue, collision) -> currentValue.registerCollision(collision.getOnlyRemoteName()));
+                }
+                return mapping;
+            });
+
+            return Optional.ofNullable(datasetMap.get(datasetName));
         }
-
-        if (!mapping.containsKey(datasetName)) {
-            // dataset doesn't exist
-            mapping.put(datasetName, Optional.empty());
+        catch (ExecutionException e) {
```

Comment:
What throws these, the cache? When writing or reading? Maybe leave a comment.

Reply:
`com.google.common.cache.Cache#get` may throw the exception. I feel the comment isn't required as it's a check exception.